spring-integration-zookeeper/src/main/java/org/springframework/integration/zookeeper/metadata/ZookeeperMetadataStore.java

```diff
@@ -0,0 +1,381 @@
+
+import org.springframework.context.SmartLifecycle;
+import org.springframework.util.Assert;
+
+/**
+ * Zookeeper-based {@link ListenableMetadataStore} based on a Zookeeper node. Values are stored in the children node,
+ * the names of which are stored as keys.
+ *
+ * @author Marius Bogoevici
+ */
+public class ZookeeperMetadataStore implements ListenableMetadataStore, SmartLifecycle {
+
+	private final Object lifecycleMonitor = new Object();
+
+	private final CuratorFramework client;
+
+	private final List<MetadataStoreListener> listeners = new CopyOnWriteArrayList<>();
+
+	/**
+	 * An internal map storing local updates, ensuring that they have precedence if the cache contains stale data.
+	 * As changes are propagated back from Zookeeper to the cache, entries are removed.
+	 */
+	private final ConcurrentMap<String, LocalChildData> updateMap = new ConcurrentHashMap<>();
+
+	private volatile String root = "/SpringIntegration-MetadataStore";
+
+	private volatile String encoding = "UTF-8";
+
+	private volatile PathChildrenCache cache;
+
+	private volatile boolean running = false;
+
+	private volatile boolean autoStartup = false;
+
+	private volatile int phase = Integer.MAX_VALUE;
+
+	public ZookeeperMetadataStore(CuratorFramework client) throws Exception {
+		Assert.notNull(client, "Client cannot be null");
+		this.client = client;
+	}
+
+	/**
+	 * Encoding to use when storing data in ZooKeeper
+	 *
+	 * @param encoding encoding as text
+	 */
+	public void setEncoding(String encoding) {
+		Assert.notNull(encoding, "'encoding' must not be null.");
+		this.encoding = encoding;
+	}
+
+	/**
+	 * Root node - store entries are children of this node
+	 *
+	 * @param root encoding as text
+	 */
+	public void setRoot(String root) {
+		Assert.notNull(root, "'root' must not be null.");
+		Assert.isTrue(root.startsWith("/"), "'root' must start with '/'");
+		Assert.isTrue(!root.endsWith("/"), "'root' must not end with '/'");
+		this.root = root;
+	}
+
+	public String getRoot() {
+		return this.root;
+	}
+
+	public void setAutoStartup(boolean autoStartup) {
+		this.autoStartup = autoStartup;
+	}
+
+	public void setPhase(int phase) {
+		this.phase = phase;
+	}
+
+	@Override
+	public String putIfAbsent(String key, String value) {
+		Assert.notNull(key, "'key' must not be null.");
+		Assert.notNull(value, "'value' must not be null.");
+		synchronized (this.updateMap) {
+			try {
+				createNode(key, value);
+				return null;
+			}
+			catch (KeeperException.NodeExistsException e) {
+				// so the data actually exists, we can read it
+				try {
+					byte[] bytes = this.client.getData().forPath(getPath(key));
+					return Conversions.bytesToString(bytes, "UTF-8");
+				}
+				catch (Exception exceptionDuringGet) {
+					throw new ZookeeperMetadataStoreException("Exception while creating node with key '" + key + "':", e);
+				}
+			}
+			catch (Exception e) {
+				throw new ZookeeperMetadataStoreException("Error while trying to set '" + key + "':", e);
```

Comment:
Well, looks like we should have a generic `MetadataStoreException` instead...
But we can live with this for now as well.